File: oc/UITextFieldDemo/UITextFieldDemo/ViewController.m
//
//  ViewController.m
//  UITextFieldDemo
//
//  Created by dfang on 2018-1-15.
//  Copyright © 2018年 east. All rights reserved.
//

#import "ViewController.h"
#import "KeyboardToolbar.h"

@interface ViewController () <KeyboardToolbarDelegate>

@property (weak, nonatomic) IBOutlet UIView *fieldsContainerView;

@property (nonatomic, strong) KeyboardToolbar *toolbar;

@property (nonatomic, strong) NSArray *fields;

@end

@implementation ViewController

- (void)viewDidLoad {
    [super viewDidLoad];
    // Do any additional setup after loading the view, typically from a nib.
    
    NSMutableArray *fieldsM = [NSMutableArray array];
    for (UIView *subView in self.fieldsContainerView.subviews) {
        if ([subView isKindOfClass:[UITextField class]]) {
            UITextField *tf = (UITextField *)subView;
            tf.inputAccessoryView = self.toolbar;
            [fieldsM addObject:tf];
        }
    }
    _fields = fieldsM;
    
    [[NSNotificationCenter defaultCenter] addObserver:self selector:@selector(keyboardWillShow:) name:UIKeyboardWillShowNotification object:nil];
    [[NSNotificationCenter defaultCenter] addObserver:self selector:@selector(keyboardWillHidden:) name:UIKeyboardWillHideNotification object:nil];
}

- (KeyboardToolbar *)toolbar {
    if (_toolbar == nil) {
        _toolbar = [KeyboardToolbar toolbar];
        _toolbar.keyboardDelegate = self;
    }
    
    return _toolbar;
}

- (void)keyboardWillShow: (NSNotification *)notify {
    UITextField *curTF = [self currentFirstResponderField];
    CGRect keyboardFrame = [notify.userInfo[UIKeyboardFrameEndUserInfoKey] CGRectValue];
    
    CGFloat deltaY = (self.fieldsContainerView.frame.origin.y + CGRectGetMaxY(curTF.frame)) - keyboardFrame.origin.y;
    
    NSInteger index = [self.fields indexOfObject:curTF];
    self.toolbar.previousItem.enabled = (index > 0);
    self.toolbar.nextItem.enabled = (index < self.fields.count - 1);
    
    if (deltaY > 0) {
        [UIView animateWithDuration:0.25 animations:^{
            self.view.transform = CGAffineTransformMakeTranslation(0, -deltaY);
        }];
    }
}

- (void)keyboardWillHidden: (NSNotification *)notify {
    [UIView animateWithDuration:0.25 animations:^{
        self.view.transform = CGAffineTransformIdentity;
        self.toolbar.previousItem.enabled = YES;
        self.toolbar.nextItem.enabled = YES;
    }];
}

// KeyboardToolbarDelegate
- (void)keyboardClickedWithItem:(UIBarButtonItem *)item {
    UITextField *curTF = [self currentFirstResponderField];
    NSInteger index = [self.fields indexOfObject:curTF];
    
    switch (item.tag) {
        case 0: // 上一个
        {
            if (index > 0) {
                [curTF resignFirstResponder];
                [self.fields[index - 1] becomeFirstResponder];
            }
        }
            break;
        case 1: // 下一个
        {
            if (index < self.fields.count) {
                [curTF resignFirstResponder];
                [self.fields[index + 1] becomeFirstResponder];
            }
        }
            break;
        case 2: // Done
            [self.fieldsContainerView endEditing:true];
            break;
        default:
            break;
    }
}

- (UITextField *)currentFirstResponderField {
    for (UITextField *tf in self.fields) {
        if (tf.isFirstResponder) {
            return tf;
        }
    }
    return nil;
}

@end

File: swift/UITextFieldDemo/UITextFieldDemo/ViewController.swift
//
//  ViewController.swift
//  UITextFieldDemo
//
//  Created by dfang on 2018-1-12.
//  Copyright © 2018年 east. All rights reserved.
//

import UIKit

class ViewController: UIViewController, KeyboardToolbarDelegate {
    @IBOutlet weak var fieldsContainerView: UIView!
    
    lazy var keyboardToolbar: KeyboardToolbar = {
        let keyboardToolbar = KeyboardToolbar.init(frame: CGRect.init(x: 0, y: 0, width: self.view.frame.size.width, height: 40))
        return keyboardToolbar
    }()
    
    var fields:Array<UITextField> = []
    
    override func viewDidLoad() {
        super.viewDidLoad()
        
        keyboardToolbar.keyboardDelegate = self
        
        for subview in fieldsContainerView.subviews {
            if (subview.isKind(of: UITextField.self)) {
                let subTextField = subview as! UITextField
                subTextField.inputAccessoryView = keyboardToolbar
                fields.append(subTextField)
            }
        }
        
        // add notification for keyboard
        NotificationCenter.default.addObserver(self, selector: #selector(keyboardWillShow(notify:)), name: NSNotification.Name.UIKeyboardWillShow, object: nil)
        NotificationCenter.default.addObserver(self, selector: #selector(keyboardWillHidden(notify:)), name: NSNotification.Name.UIKeyboardWillHide, object: nil)
    }

    @objc func keyboardWillShow(notify: NSNotification) {
        guard let curTf = firstResponderTextField() else {
            return
        }
        guard let index = fields.index(of: curTf) else { return }
        keyboardToolbar.previousBtn.isEnabled = (index > 0)
        keyboardToolbar.nextBtn.isEnabled = (index < fields.count - 1)
        
        let userInfo = notify.userInfo!
        let keyboardFrame = (userInfo[UIKeyboardFrameEndUserInfoKey] as! NSValue).cgRectValue
        let duration = (userInfo[UIKeyboardAnimationDurationUserInfoKey] as! NSNumber).doubleValue
        let deltaY = (fieldsContainerView.frame.origin.y + curTf.frame.maxY) - keyboardFrame.origin.y
        
        if (deltaY > 0) { // keyboard cover the first responder uitextfield
            UIView.animate(withDuration: duration, animations: {
                self.view.transform = CGAffineTransform.init(translationX: 0, y: -deltaY)
            })
        }
    }
    
    @objc func keyboardWillHidden(notify: NSNotification) {
        let duration = (notify.userInfo![UIKeyboardAnimationDurationUserInfoKey] as! NSNumber).doubleValue
        UIView.animate(withDuration: duration) {
            self.view.transform = CGAffineTransform.identity
        }
        
        // 还原工具条的item为可用
        keyboardToolbar.previousBtn.isEnabled = true
        keyboardToolbar.nextBtn.isEnabled = true
    }
    
    // KeyboardToolbarDelegate
    func btnDidClicked(btn: UIBarButtonItem) {
        guard let firstTf = firstResponderTextField() else { return }
        
        guard var index = fields.index(of: firstTf) else { return }
        
        switch btn.title! {
        case "上一个":
            if (index > 0) {
                index = index - 1
                firstTf.resignFirstResponder()
                fields[index].becomeFirstResponder()
            }
            break
        case "下一个":
            if (index < fields.count - 1) {
                index = index + 1
                firstTf.resignFirstResponder()
                fields[index].becomeFirstResponder()
            }
            break
        case "Done":
            fieldsContainerView.endEditing(true)
            break
        default:
            break
        }
    }
    
    func firstResponderTextField() -> UITextField? {
        for subTextField:UITextField in fields {
            if (subTextField.isFirstResponder) {
                return subTextField
            }
        }
        return nil
    }
    
    override func touchesBegan(_ touches: Set<UITouch>, with event: UIEvent?) {
        fieldsContainerView.endEditing(true)
    }
}

